group/assignment6/RearrangeCars.cpp

```diff
@@ -35,36 +31,29 @@ void RearrangeCars::moveCar(vector<int> &currentStates, vector<int> &carPosition
 	carPosition[currentStates[to]] = to;
 }
 
-void RearrangeCars::generateSequenceOfMoves(const vector<int> &endStates, vector<Move> &sequenceOfMoves) {
-	vector<int> currentStates = this->startStates; // startStates will remain unchanged
-	vector<int> carPosition = findCarPositions(currentStates);
-
-	int emptySlotId = findEmptySlotId();
-	int currCarId, currCarSlotId;
-
-	while (true) {
-		currCarId = endStates[emptySlotId];
-		if (currCarId == -1)	// We reached slot which should be empty in the end
-			currCarId = findMisplacedCar(carPosition, endStates);
-		if (currCarId == -1) break;
-		currCarSlotId = carPosition[currCarId];
-		moveCar(currentStates, carPosition, currCarSlotId, emptySlotId);
-		sequenceOfMoves.push_back(Move(currCarId, make_pair(currCarSlotId, emptySlotId)));
-		emptySlotId = currCarSlotId;
-	}
-}
+bool RearrangeCars::validMove(int carId, int emptySlotId, const vector<set<int>>* constrains) {
+	if (!constrains)	// Constrains are not set
+		return true;
 
-bool RearrangeCars::validMove(int carId, int emptySlotId, const vector<set<int>>& constrains) {
 	if (constrains[emptySlotId].size() == 0)	// There are no constraints for this parking space, any car can be moved here
 		return true;
 
-	if ((constrains[emptySlotId]).find(carId) == (constrains[emptySlotId].end()))
+	if (((*constrains)[emptySlotId]).find(carId) == ((*constrains)[emptySlotId].end()))
 		return true;
 
 	return false;
 }
 
-int RearrangeCars::findMisplacedCarWithConstraints(vector<int> &carPosition, vector<int> &currentStates, const vector<int> &endStates, const vector<set<int>> &constraints, vector<Move> &sequenceOfMoves, int &emptySlotId, int &numOfCarsOnDisiredPosition) {
+int RearrangeCars::numberOfCarsOnDesiredPosition(const vector<int>& carPosition, const vector<int>& endStates) {
+	int count = 0;
+	for (int i = 0; i < carPosition.size(); i++)
+		if (endStates[carPosition[i]] == i)
+			count++;
+
+	return count;
+}
+
+int RearrangeCars::findMisplacedCar(vector<int> &carPosition, vector<int> &currentStates, const vector<int> &endStates, const vector<set<int>> *constraints, vector<Move> &sequenceOfMoves, int &emptySlotId, int &numOfCarsOnDisiredPosition) {

```

Comment:
when you have too many parameters put them under each other to make it easier for reader. But then I searched what is appropriate way to pass many parameters and found out that we should make a structure struct. For example, parkingLot and it will have fields: carPosition, currentStates, endStates, constraints, emptySlotId and numOfCarsOnDisiredPosition. so we pass only that structure object and sequenceOfMoves to function. bruteForceAllSequence function can also use this structure. What you think?

Reply:
Also this function does not only finds misplaced car, so name must be changed.

Reply:
I think this can be postponed to the next iteration so that we can further discuss if we need a struct.